File: algorithms/1105.uncrossed_lines/uncrossed_lines.cpp
class Solution {
public:
    int maxUncrossedLines(vector<int>& nums1, vector<int>& nums2) {
        const size_t n = nums1.size(), m = nums2.size();
        vector<int> dp(m + 1, 0);
        vector<int> last(m + 1, 0);
        for (size_t i = 1; i <= n; ++i) {
            for (size_t j = 1; j <= m; ++j) {
                if (nums1[i - 1] == nums2[j - 1]) {
                    dp[j] = last[j - 1] + 1;
                } else {
                    dp[j] = max(last[j], dp[j - 1]);
                }
            }
            dp.swap(last);
        }
        return max(dp[m], last[m]);
    }
};

File: algorithms/1105.uncrossed_lines/uncrossed_lines.rs
impl Solution {
    pub fn max_uncrossed_lines(nums1: Vec<i32>, nums2: Vec<i32>) -> i32 {
        let (n, m) = (nums1.len(), nums2.len());
        let mut dp = vec![0; m + 1];
        let mut last = vec![0; m + 1];
        for i in 1..=n {
            for j in 1..=m {
                if nums1[i - 1] == nums2[j - 1] {
                    dp[j] = last[j - 1] + 1;
                } else {
                    dp[j] = last[j].max(dp[j - 1]);
                }
            }
            dp.swap_with_slice(&mut last)
        }
        last[m].max(dp[m])
    }
}